### ISSUE TITLE
PmagPy 4.0

### ISSUE THREAD

**Issue by moonshoes87** [PmagPy/PmagPy#405]
@ltauxe, I believe everything all of the conversion scripts now: are updated to 3.0, moved into convert_2_magic, and have working tests!  

Next tasks:

- add all conversion scripts to _PmagPy-cli.ipynb
- improve and add doc-strings to all convert_2_magic functions that are currently missing them, in the standard numpy format #138.  
- drag the last few programs into 3.0 that are still stuck in 2.5

Of those tasks, I believe only updating programs from 3.0 to 2.5 should be done before the pip release.  So, the 3.0 --> 4.0 transition means: we've finished upgrading to MagIC 3 (for real this time), and we've changed the API for interacting with conversion scripts.  The conversion scripts in particular have breaking changes, so it's a useful signal to people that we are jumping a version number.

I believe these are the only two remaining programs that need to be upgraded to MagIC 3:

revtest_magic #374 
chi_magic #365 

I will also check and see if there are any more that have been missed.  So, my plan is to get those upgraded, make sure all tests are passing and notebooks run, and have you do a pip release.  We can continue working on the task list above after the 4.0 release.

**Comment by moonshoes87** [comment]
@ltauxe, now would also be a good time to rename any functions or whatever that you don't like.  For instance, I think we've talked about possibly renaming some of the functions in pmagplotlib.py to PEP 8 style guidelines?  I have a script that can go in and replace a name throughout the repo, so it is pretty simple, we just have to decide to do it.

**Comment by ltauxe** [comment]
good idea.  there are a few in pmag.py but quite a lot in pmagplotlib.py   won’t that break lots of stuff?  or is it worth it?  
lisa

**Comment by moonshoes87** [comment]
@ltauxe, Unfortunately, here is a list of sneaky programs that *still* aren't in 3.0.

Can I deprecate:

change\_case\_magic.py?

and:

site\_edit\_magic.py?

The rest of these either need to be upgraded to 3 (and possibly moved into conversion_scripts), or deprecated.  Let me know what you want to happen with each of them.  I really thought there weren't this many left.... sigh.
<img width="1215" alt="screen shot 2018-08-02 at 9 30 01 pm" src="https://user-images.githubusercontent.com/1836578/43624301-4896d3fe-969b-11e8-91d1-2d906a794193.png">

**Comment by moonshoes87** [comment]
As far as renaming is concerned, my magic auto-replace script (using sed) will replace all instances of a string in the repo.  So, if we want to rename drawFIGS to draw_figs, it will replace it everywhere it is used.  It will only break integrations if people are using pmagplotlib in their own code -- but that is to be expected with jumping a version #.

**Comment by moonshoes87** [comment]
@ltauxe, if you can weigh in on which of these things you think should happen or not, and what if anything else you want to add for the 4.0 release. 

name changes:  

- new_builder.py to contribution_builder.py.
- some of the pmagplotlib functions to pep 8 standards

deprecate:

- change_case_magic
- maybe grab_magic_key
- possibly some others from the list above?

update to MagIC 3:

- some more of the listed programs???

**Comment by ltauxe** [comment]
yes let's move new_builder.py to contribution_builder.py 
let's change all the pmagplotlib functions to pep 8 standards AND ADD PROPER DOC STRING

deprecate:  change_case_magic.
i still use grab_magic_key on occasion. 
keep site_edit_magic and make it 3.0  

chi_magic can be upgraded to 3.0 - see PmagPy_nb.  
i think revtest_magic may already be upgraded in ipmag (by another name).

**Comment by ltauxe** [comment]
there is only one exception to this that i know of - i put a cartopy version of plotMAP called plot_map  - i’ll change that to plot_map_cartopy so we won’t get snarled up.  
lisa

**Comment by ltauxe** [comment]
well, let’s hold our breath and give it a go.

**Comment by moonshoes87** [comment]
chi_magic and revtest_magic are already upgraded (and not on the list ;) )

I will deprecate: change_case_magic

I will upgrade: site_edit_magic, grab_magic_key

I need to know about: scalc_magic, umich_magic, biplot_magic, microwave_magic, site_edit_magic, trmaq_magic, plotxy_magic, magic_select, susar4_asc_magic, plot_magic_keys

I will change function names in pmagplotlib, and open an issue for making docstrings (since those are not breaking changes, they can be added after we go to MagIC 4).  I'm renaming new_builder right now (that's the annoying one).

**Comment by ltauxe** [comment]
deprecate:
	plotxy_magic
	plot_magic_keys
	umich_magic (they closed up shop so no new data will be coming)
	biplot_magic
	susar4_asc_magic -  i might bring this back someday, but I don’t know who uses it - Nick Jarboe?   

i think grab_magic_key and magic_select are data model agnostic.  I find them handy on the command line so don’t deprecate them

scalc_magic should be upgraded to data model 3:  see notebook version though.  
same for trmaq_magic.   

microwave_magic was for the liverpool data format - Nick Jarboe?

**Comment by moonshoes87** [comment]
Ok, I will take a break from the grand pmagplotlib renaming and look at upgrading trmaq_magic and scalc_magic

**Comment by ltauxe** [comment]
no these are low priority.  really.  
lisa

**Comment by moonshoes87** [comment]
ok, I'll just put them on the list.  what about convert_samples_magic?

**Comment by ltauxe** [comment]
all that does is takes a samples table and creates a template for the orient.txt file.  I think that is what step 2 in pmag_gui does pretty much anyway - go ahead and deprecate.  
lisa

**Comment by moonshoes87** [comment]
ok, I finished with pmagplotlib.  There are a lot of functions in there!  I added a note #not used by all the functions that aren't used at all in the code (including in any notebook) -- there are about 10 of them.  @ltauxe, can I remove them or do you want to have a look at them first?

Also, I tried in a few cases to rename the functions for clarity.  For example, plotA --> plot_arai.  I was hoping to rename plotAZ to something more clear, but I couldn't figure out what.  Any thoughts?

**Comment by moonshoes87** [comment]
Some more name changes for clarity:

- pmag_basic_dialogs_native3 --> pmag_gui_dialogs
- pmag_menu_dialogs --> pmag_gui_menu_dialogs
- pmag_er_magic_dialogs --> ???  
- pmag_basic_dialogs --> pmag_gui_dialogs2  ??

**Comment by moonshoes87** [comment]
Ok, I think I've done all the things I wanted to do, except these items for input from @ltauxe :

- input on renaming pmagplotlib.plotAZ
- input on removing the 10 pmagplotlib functions that are never used
- are there other breaking changes to make before 4.0?  I.e., renaming things, deprecating things, etc.

After that, I will make sure all tests are passing, and notebooks are running, and then you can do the pip release @ltauxe .

**Comment by ltauxe** [comment]
i’ll take a look soon but right now can’t. sorry.

**Comment by ltauxe** [comment]
PlotAZ plots Arai and Zijderveld plots.ill take a look soon

**Comment by moonshoes87** [comment]
Can we change plotAZ to plot_arai_zij?

Can I remove the unused pmagplotlib functions, or do you want to just leave them with the 'deprecated' docstring?  I think we should remove them, since we can always bring them back if needed.

**Comment by ltauxe** [comment]
i just changed it to plot_az, but plot_arai_zij would be better.  and yes you can just remove them.  they are very very old.

**Comment by moonshoes87** [comment]
great, I will do that name change, and remove the deprecated functions.

of the currently open issues, which (if any) do you want me to crush before we make the 4.0 pip release?

**Comment by ltauxe** [comment]
i think all the other things on my list are minor with the possilble exception of aarm_magic and atrm_magic.  could you maybe make those ipmag version like all the others?  then I can fiddle with them in the notebook.  after that, we can go to 4.0  and clean up the cookbook!  
lisa

**Comment by moonshoes87** [comment]
ok, I'll work on aarm_magic and atrm_magic next!

**Comment by ltauxe** [comment]
Ok how abt tomorrow afternoon
